nvme.c

```diff
@@ -159,6 +162,22 @@ void *nvme_alloc(size_t len, bool *huge)
 }
 #endif
 
+static int getrandom_bytes(void *buf, size_t buflen)
+{
+#if HAVE_SYS_RANDOM
+	return getrandom(buf, buflen, GRND_NONBLOCK);
+#else
+	int result = 0;
+	int fd = open("/dev/urandom", O_RDONLY);
+	if (fd < 0)
+		return fd;
+	else
+		result = read(fd, buf, buflen);
+	close(fd);
+	return result;
```

Comment:
If you don't mind I would change the above to something like this here:
```
	int fd, err = 0;

	fd = open("/dev/urandom", O_RDONLY);
	if (fd < 0)
		return fd;
	if (read(fd, buf, buflen) < 0)
		err = errno;
	close(fd);
	errno = err;
	return err? -1 : 0;
```
The close will overwrite the errno from the read. Hence you need to cache it.

Reply:
I would like to remain my code,
it should be works same with exist sys/random.h , so return value should be read length.

getrandom 
 On success, getrandom() returns the number of bytes
 On error, -1 is returned, and errno is set to indicate the error.

read
 On success, the number of bytes read is returned 
 On error, -1 is returned, and errno is set to indicate the error.

Reply:
Good point. But let me still add the errno caching.

Reply:
In this case let's get also the return type correct :)

```
static ssize_t getrandom_bytes(void *buf, size_t buflen)
{
#if HAVE_SYS_RANDOM
	return getrandom(buf, buflen, GRND_NONBLOCK);
#else
	ssize_t result;
	int fd, err = 0;

	fd = open("/dev/urandom", O_RDONLY);
	if (fd < 0)
		return fd;
	result = read(fd, buf, buflen);
	if (result < 0)
		err = errno;
	close(fd);
	errno = err;
	return result;
#endif
}```

Would that work for you?

Reply:
@igaw great :)